azurerm/internal/services/iottimeseriesinsights/resource_iot_time_series_insights_access_policy.go

```diff
@@ -0,0 +1,212 @@
+			return tf.ImportAsExistsError("azurerm_iot_time_series_insights_access_policy", *existing.ID)
+		}
+	}
+
+	policy := timeseriesinsights.AccessPolicyCreateOrUpdateParameters{
+		&timeseriesinsights.AccessPolicyResourceProperties{
+			Description:       utils.String(d.Get("description").(string)),
+			PrincipalObjectID: utils.String(d.Get("principal_object_id").(string)),
+			Roles:             expandIoTTimeSeriesInsightsAccessPolicyRoles(d.Get("roles").(*schema.Set).List()),
+		},
+	}
+
+	_, err := client.CreateOrUpdate(ctx, resourceGroup, environmentName, name, policy)
+	if err != nil {
+		return fmt.Errorf("creating/updating IoT Time Series Insights Access Policy %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, environmentName, name)
+	if err != nil {
+		return fmt.Errorf("retrieving IoT Time Series Insights Access Policy %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("cannot read IoT Time Series Insights Access Policy %q (Resource Group %q) ID", name, resourceGroup)
+	}
+
+	d.SetId(*resp.ID)
+
+	return resourceArmIoTTimeSeriesInsightsAccessPolicyRead(d, meta)
+}
+
+func resourceArmIoTTimeSeriesInsightsAccessPolicyRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).IoTTimeSeriesInsights.AccessPoliciesClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.TimeSeriesInsightsAccessPolicyID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.EnvironmentName, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("retrieving IoT Time Series Insights Access Policy %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", id.ResourceGroup)
+	d.Set("environment_name", id.EnvironmentName)
+
+	if props := resp.AccessPolicyResourceProperties; props != nil {
+		d.Set("description", props.Description)
+		d.Set("principal_object_id", props.PrincipalObjectID)
+		d.Set("roles", flattenIoTTimeSeriesInsightsAccessPolicyRoles(resp.Roles))
+	}
+
+	return nil
+}
+
+func resourceArmIoTTimeSeriesInsightsAccessPolicyDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).IoTTimeSeriesInsights.AccessPoliciesClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.TimeSeriesInsightsAccessPolicyID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	response, err := client.Delete(ctx, id.ResourceGroup, id.EnvironmentName, id.Name)
+	if err != nil {
+		if !utils.ResponseWasNotFound(response) {
+			return fmt.Errorf("deleting IoT Time Series Insights Access Policy %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+		}
```

Comment:
I think we should not escape `404` errors here. Whether the SDK gives non-nil errors is defined [here](https://github.com/Azure/azure-rest-api-specs/blob/3b552e66a85a3de4cc51de5998989c3db6a63920/specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/preview/2018-08-15-preview/timeseriesinsights.json#L1026) in the `response` section of the swagger. For the listed status codes, SDK will not return non-nil errors for them (in this case, they are 200 and 204). And all other status codes, the SDK will return non-nil errors using the error schema defined [here](https://github.com/Azure/azure-rest-api-specs/blob/3b552e66a85a3de4cc51de5998989c3db6a63920/specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/preview/2018-08-15-preview/timeseriesinsights.json#L1036).

This means in this case, `404` will always be an error and indicates something is wrong. If this resource does not exist, and we still want to delete this, the service should return `204` for this.

One of the scenario `404` was returned in deletion should be we are deleting this resource, but the resource group does not exist. I believe in this case the service would return `404`. But in this case I believe we should call out and throw errors.
What do you think?